rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -355,62 +354,64 @@ object Reduce {
       * @param neu
       * @return
       */
-    def eval(neu: New)(implicit env: Env[Par]): M[Unit] = {
+    private def eval(neu: New)(implicit env: Env[Par]): M[Unit] = {
       def alloc(level: Int): Env[Par] =
         (env /: (0 until level).toList) { (_env, _) =>
           val addr: Par = GPrivate()
           _env.put(addr)
         }
 
-      eval(neu.p.get)(alloc(neu.bindCount))
+      // TODO: should cost of `new` construct depend on the # of bindings?
+      // they will be cost further in the reduction
+      eval(neu.p)(alloc(neu.bindCount)) <* costAccounting.modify(_ + NEW_EVAL_COST)
     }
 
     private[this] def unbundleReceive(rb: ReceiveBind)(implicit env: Env[Par]): M[Quote] =
       for {
-        quote <- eval(rb.source.get)
+        quote <- eval(rb.source)
         subst <- substituteQuote[M].substitute(quote)(0, env)
         // Check if we try to read from bundled channel
         unbndl <- subst.quote.get.singleBundle() match {
                    case Some(value) =>
                      if (!value.readFlag) {
-                       interpreterErrorM[M].raiseError(
-                         ReduceError("Trying to read from non-readable channel."))
+                       s.raiseError(ReduceError("Trying to read from non-readable channel."))
                      } else {
-                       interpreterErrorM[M].pure(Quote(value.body.get))
+                       s.pure(Quote(value.body))
                      }
                    case None =>
-                     interpreterErrorM[M].pure(subst)
+                     s.pure(subst)
                  }
+        _ <- costAccounting.modify(_ + UNBUNDLE_RECEIVE_COST)
       } yield unbndl
 
-    def eval(bundle: Bundle)(implicit env: Env[Par]): M[Unit] =
-      eval(bundle.body.get)
+    private def eval(bundle: Bundle)(implicit env: Env[Par]): M[Unit] =
+      eval(bundle.body)
 
     def evalExprToPar(expr: Expr)(implicit env: Env[Par]): M[Par] =
       expr.exprInstance match {
         case EVarBody(EVar(v)) =>
           for {
-            p       <- eval(v.get)
+            p       <- eval(v)
             evaledP <- evalExpr(p)
           } yield evaledP
         case EMethodBody(EMethod(method, target, arguments, _, _)) => {
           val methodLookup = methodTable(method)
           for {
-            evaledTarget <- evalExpr(target.get)
+            _            <- costAccounting.modify(_ + METHOD_CALL_COST)
```

Comment:
Different methods might have different costs, but this is fine for now.

Reply:
I agree, I planned to put method specific costs into the evaluation itself and that was supposed to be constant price of the method call.